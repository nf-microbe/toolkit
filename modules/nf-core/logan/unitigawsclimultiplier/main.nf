process LOGAN_UNITIGAWSCLIMULTIPLIER {
    tag "${meta.id}"
    label 'process_high'

    conda "${moduleDir}/environment.yml"
    container "${ workflow.containerEngine == 'singularity' && !task.ext.singularity_pull_docker_container ?
        'oras://community.wave.seqera.io/library/seqkit_awscli_biopython:4bbd3cad980280ad' :
        'community.wave.seqera.io/library/seqkit_awscli_biopython:1ebd36d34b5de691' }"

    input:
    tuple val(meta), val(accession)

    output:
    tuple val(meta), path("${accession}.unitigs.fa.gz")         , emit: raw_fasta
    tuple val(meta), path("${accession}.filter.fasta.gz")       , emit: filtered_fasta      , optional: true
    tuple val(meta), path("${accession}.multiplier.fasta.gz")   , emit: multiplied_fasta    , optional: true
    tuple val(meta), path("${accession}.filt_mult.fasta.gz")    , emit: filt_mult_fasta     , optional: true
    path "versions.yml"                                         , emit: versions

    when:
    task.ext.when == null || task.ext.when

    script:
    def args            = task.ext.args ?: ''
    def run_filt        = task.ext.args.contains("-a")
    def run_mult        = task.ext.args.contains("-m")
    def run_filt_mult   = task.ext.args.contains("-a") && task.ext.args.contains("-m")
    prefix              = task.ext.prefix ?: "${meta.id}"
    """
    ### Download Logan unitigs
    aws s3 cp s3://logan-pub/u/${accession}/${accession}.unitigs.fa.zst . --no-sign-request || true

    ### Multiply Logan unitigs based on average kmer abundance
    zstd -d ${accession}.unitigs.fa.zst
    rm ${accession}.unitigs.fa.zst

    multiplier.py \\
        -i ${accession}.unitigs.fa \\
        -o ${accession} \\
        ${args}

    gzip ${accession}.unitigs.fa

    if [ "${run_filt}" == "true" ]; then
        gzip ${accession}.filter.fasta
    fi

    if [ "${run_mult}" == "true" ]; then
        gzip ${accession}.multiplier.fasta
    fi

    if [ "${run_filt_mult}" == "true" ]; then
        gzip ${accession}.filt_mult.fasta
    fi

    cat <<-END_VERSIONS > versions.yml
    "${task.process}":
        awscli: \$( aws --version | sed 's/aws-cli\\///; s/ Python.*//' )
        seqkit: \$(seqkit version | cut -d' ' -f2)
        python: \$( python --version | sed 's/Python //' )
        biopython: \$(python -c "import Bio; print(Bio.__version__)")
    END_VERSIONS
    """

    stub:
    def args            = task.ext.args ?: ''
    def run_filt        = task.ext.args.contains("-a")
    def run_mult        = task.ext.args.contains("-m")
    def run_filt_mult   = task.ext.args.contains("-a") && task.ext.args.contains("-m")
    prefix              = task.ext.prefix ?: "${meta.id}"
    """
    echo "" | gzip > ${accession}.unitigs.fa.gz

    if [ "${run_filt}" == "true" ]; then
        echo "" | gzip > ${accession}.filtered.fasta.gz
    fi

    if [ "${run_mult}" == "true" ]; then
        echo "" | gzip > ${accession}.multiplier.fasta.gz
    fi

    if [ "${run_filt_mult}" == "true" ]; then
        echo "" | gzip > ${accession}.filt_mult.fasta.gz
    fi

    cat <<-END_VERSIONS > versions.yml
    "${task.process}":
        awscli: \$( aws --version | sed 's/aws-cli\\///; s/ Python.*//' )
        seqkit: \$(seqkit version | cut -d' ' -f2)
        python: \$( python --version | sed 's/Python //' )
        biopython: \$(python -c "import Bio; print(Bio.__version__)")
    END_VERSIONS
    """
}
