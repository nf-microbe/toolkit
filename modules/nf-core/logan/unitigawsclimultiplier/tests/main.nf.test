nextflow_process {

    name "Test Process LOGAN_UNITIGAWSCLIMULTIPLIER"
    script "../main.nf"
    process "LOGAN_UNITIGAWSCLIMULTIPLIER"
    config "./nextflow.config"

    tag "modules"
    tag "modules_nfcore"
    tag "logan"
    tag "logan/unitigawsclimultiplier"

    test("SRR24085752 - raw only") {
        when {
            params {
                ext_args = ''
            }
            process {
                """
                input[0] = [
                    [ id:'SRR24085752', single_end:false, md5_1: 'e67296355980997dde170c602d9cb4c1', md5_2: '86b4ee5d4860aed0a3a2538a7757cbc8' ], // meta map
                    'SRR24085752'
                ]
                """
            }
        }
        then {
            assertAll (
                { assert process.success },
                { assert snapshot(process.out).match() }
            )
        }
    }

    test("SRR24085752 - filtered") {
        when {
            params {
                ext_args = '--abundance 5'
            }
            process {
                """
                input[0] = [
                    [ id:'SRR24085752', single_end:false, md5_1: 'e67296355980997dde170c602d9cb4c1', md5_2: '86b4ee5d4860aed0a3a2538a7757cbc8' ], // meta map
                    'SRR24085752'
                ]
                """
            }
        }
        then {
            assertAll (
                { assert process.success },
                { assert snapshot(process.out).match() }
            )
        }
    }

    test("SRR24085752 - multiplied") {
        when {
            params {
                ext_args = '--multiplier'
            }
            process {
                """
                input[0] = [
                    [ id:'SRR24085752', single_end:false, md5_1: 'e67296355980997dde170c602d9cb4c1', md5_2: '86b4ee5d4860aed0a3a2538a7757cbc8' ], // meta map
                    'SRR24085752'
                ]
                """
            }
        }
        then {
            assertAll (
                { assert process.success },
                { assert snapshot(process.out).match() }
            )
        }
    }

    test("SRR24085752 - filtered + multiplied") {
        when {
            params {
                ext_args = '--multiplier --abundance 5'
            }
            process {
                """
                input[0] = [
                    [ id:'SRR24085752', single_end:false, md5_1: 'e67296355980997dde170c602d9cb4c1', md5_2: '86b4ee5d4860aed0a3a2538a7757cbc8' ], // meta map
                    'SRR24085752'
                ]
                """
            }
        }
        then {
            assertAll (
                { assert process.success },
                { assert snapshot(process.out).match() }
            )
        }
    }

    test("stub - raw") {
        options "-stub"
        when {
            params {
                ext_args = ''
            }
            process {
                """
                input[0] = [
                    [ id:'SRR24085752', single_end:false, md5_1: 'e67296355980997dde170c602d9cb4c1', md5_2: '86b4ee5d4860aed0a3a2538a7757cbc8' ], // meta map
                    'SRR24085752'
                ]
                """
            }
        }
        then {
            assertAll (
                { assert process.success },
                { assert snapshot(process.out).match() }
            )
        }
    }

    test("stub - filtered + multiplied") {
        options "-stub"
        when {
            params {
                ext_args = '--multiplier --abundance 5'
            }
            process {
                """
                input[0] = [
                    [ id:'SRR24085752', single_end:false, md5_1: 'e67296355980997dde170c602d9cb4c1', md5_2: '86b4ee5d4860aed0a3a2538a7757cbc8' ], // meta map
                    'SRR24085752'
                ]
                """
            }
        }
        then {
            assertAll (
                { assert process.success },
                { assert snapshot(process.out).match() }
            )
        }
    }
}
