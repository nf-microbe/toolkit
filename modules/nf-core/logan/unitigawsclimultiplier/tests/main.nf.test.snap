{
    "SRR24085752 - raw only": {
        "content": [
            {
                "0": [
                    [
                        {
                            "id": "SRR24085752",
                            "single_end": false,
                            "md5_1": "e67296355980997dde170c602d9cb4c1",
                            "md5_2": "86b4ee5d4860aed0a3a2538a7757cbc8"
                        },
                        "SRR24085752.unitigs.fa.gz:md5,134bb7521e3af1d7cf0ceb3c465bdc53"
                    ]
                ],
                "1": [
                    
                ],
                "2": [
                    
                ],
                "3": [
                    
                ],
                "4": [
                    "versions.yml:md5,63c274313e2b305b08175b44dea333a7"
                ],
                "filt_mult_fasta": [
                    
                ],
                "filtered_fasta": [
                    
                ],
                "multiplied_fasta": [
                    
                ],
                "raw_fasta": [
                    [
                        {
                            "id": "SRR24085752",
                            "single_end": false,
                            "md5_1": "e67296355980997dde170c602d9cb4c1",
                            "md5_2": "86b4ee5d4860aed0a3a2538a7757cbc8"
                        },
                        "SRR24085752.unitigs.fa.gz:md5,134bb7521e3af1d7cf0ceb3c465bdc53"
                    ]
                ],
                "versions": [
                    "versions.yml:md5,63c274313e2b305b08175b44dea333a7"
                ]
            }
        ],
        "meta": {
            "nf-test": "0.9.0",
            "nextflow": "24.04.4"
        },
        "timestamp": "2024-10-10T12:20:34.023879474"
    },
    "stub - raw": {
        "content": [
            {
                "0": [
                    [
                        {
                            "id": "SRR24085752",
                            "single_end": false,
                            "md5_1": "e67296355980997dde170c602d9cb4c1",
                            "md5_2": "86b4ee5d4860aed0a3a2538a7757cbc8"
                        },
                        "SRR24085752.unitigs.fa.gz:md5,68b329da9893e34099c7d8ad5cb9c940"
                    ]
                ],
                "1": [
                    
                ],
                "2": [
                    
                ],
                "3": [
                    
                ],
                "4": [
                    "versions.yml:md5,63c274313e2b305b08175b44dea333a7"
                ],
                "filt_mult_fasta": [
                    
                ],
                "filtered_fasta": [
                    
                ],
                "multiplied_fasta": [
                    
                ],
                "raw_fasta": [
                    [
                        {
                            "id": "SRR24085752",
                            "single_end": false,
                            "md5_1": "e67296355980997dde170c602d9cb4c1",
                            "md5_2": "86b4ee5d4860aed0a3a2538a7757cbc8"
                        },
                        "SRR24085752.unitigs.fa.gz:md5,68b329da9893e34099c7d8ad5cb9c940"
                    ]
                ],
                "versions": [
                    "versions.yml:md5,63c274313e2b305b08175b44dea333a7"
                ]
            }
        ],
        "meta": {
            "nf-test": "0.9.0",
            "nextflow": "24.04.4"
        },
        "timestamp": "2024-10-10T12:22:14.93459844"
    },
    "SRR24085752 - multiplied": {
        "content": [
            {
                "0": [
                    [
                        {
                            "id": "SRR24085752",
                            "single_end": false,
                            "md5_1": "e67296355980997dde170c602d9cb4c1",
                            "md5_2": "86b4ee5d4860aed0a3a2538a7757cbc8"
                        },
                        "SRR24085752.unitigs.fa.gz:md5,134bb7521e3af1d7cf0ceb3c465bdc53"
                    ]
                ],
                "1": [
                    
                ],
                "2": [
                    [
                        {
                            "id": "SRR24085752",
                            "single_end": false,
                            "md5_1": "e67296355980997dde170c602d9cb4c1",
                            "md5_2": "86b4ee5d4860aed0a3a2538a7757cbc8"
                        },
                        "SRR24085752.multiplier.fasta.gz:md5,968840dfa0f56f45294f4a389fb38297"
                    ]
                ],
                "3": [
                    
                ],
                "4": [
                    "versions.yml:md5,63c274313e2b305b08175b44dea333a7"
                ],
                "filt_mult_fasta": [
                    
                ],
                "filtered_fasta": [
                    
                ],
                "multiplied_fasta": [
                    [
                        {
                            "id": "SRR24085752",
                            "single_end": false,
                            "md5_1": "e67296355980997dde170c602d9cb4c1",
                            "md5_2": "86b4ee5d4860aed0a3a2538a7757cbc8"
                        },
                        "SRR24085752.multiplier.fasta.gz:md5,968840dfa0f56f45294f4a389fb38297"
                    ]
                ],
                "raw_fasta": [
                    [
                        {
                            "id": "SRR24085752",
                            "single_end": false,
                            "md5_1": "e67296355980997dde170c602d9cb4c1",
                            "md5_2": "86b4ee5d4860aed0a3a2538a7757cbc8"
                        },
                        "SRR24085752.unitigs.fa.gz:md5,134bb7521e3af1d7cf0ceb3c465bdc53"
                    ]
                ],
                "versions": [
                    "versions.yml:md5,63c274313e2b305b08175b44dea333a7"
                ]
            }
        ],
        "meta": {
            "nf-test": "0.9.0",
            "nextflow": "24.04.4"
        },
        "timestamp": "2024-10-10T12:21:26.743304092"
    },
    "SRR24085752 - filtered + multiplied": {
        "content": [
            {
                "0": [
                    [
                        {
                            "id": "SRR24085752",
                            "single_end": false,
                            "md5_1": "e67296355980997dde170c602d9cb4c1",
                            "md5_2": "86b4ee5d4860aed0a3a2538a7757cbc8"
                        },
                        "SRR24085752.unitigs.fa.gz:md5,134bb7521e3af1d7cf0ceb3c465bdc53"
                    ]
                ],
                "1": [
                    [
                        {
                            "id": "SRR24085752",
                            "single_end": false,
                            "md5_1": "e67296355980997dde170c602d9cb4c1",
                            "md5_2": "86b4ee5d4860aed0a3a2538a7757cbc8"
                        },
                        "SRR24085752.filter.fasta.gz:md5,5eb2cc6f1a1070e4188c72d53c3c76ae"
                    ]
                ],
                "2": [
                    [
                        {
                            "id": "SRR24085752",
                            "single_end": false,
                            "md5_1": "e67296355980997dde170c602d9cb4c1",
                            "md5_2": "86b4ee5d4860aed0a3a2538a7757cbc8"
                        },
                        "SRR24085752.multiplier.fasta.gz:md5,968840dfa0f56f45294f4a389fb38297"
                    ]
                ],
                "3": [
                    [
                        {
                            "id": "SRR24085752",
                            "single_end": false,
                            "md5_1": "e67296355980997dde170c602d9cb4c1",
                            "md5_2": "86b4ee5d4860aed0a3a2538a7757cbc8"
                        },
                        "SRR24085752.filt_mult.fasta.gz:md5,86cde1ce79975356d01ec3021c128f7a"
                    ]
                ],
                "4": [
                    "versions.yml:md5,63c274313e2b305b08175b44dea333a7"
                ],
                "filt_mult_fasta": [
                    [
                        {
                            "id": "SRR24085752",
                            "single_end": false,
                            "md5_1": "e67296355980997dde170c602d9cb4c1",
                            "md5_2": "86b4ee5d4860aed0a3a2538a7757cbc8"
                        },
                        "SRR24085752.filt_mult.fasta.gz:md5,86cde1ce79975356d01ec3021c128f7a"
                    ]
                ],
                "filtered_fasta": [
                    [
                        {
                            "id": "SRR24085752",
                            "single_end": false,
                            "md5_1": "e67296355980997dde170c602d9cb4c1",
                            "md5_2": "86b4ee5d4860aed0a3a2538a7757cbc8"
                        },
                        "SRR24085752.filter.fasta.gz:md5,5eb2cc6f1a1070e4188c72d53c3c76ae"
                    ]
                ],
                "multiplied_fasta": [
                    [
                        {
                            "id": "SRR24085752",
                            "single_end": false,
                            "md5_1": "e67296355980997dde170c602d9cb4c1",
                            "md5_2": "86b4ee5d4860aed0a3a2538a7757cbc8"
                        },
                        "SRR24085752.multiplier.fasta.gz:md5,968840dfa0f56f45294f4a389fb38297"
                    ]
                ],
                "raw_fasta": [
                    [
                        {
                            "id": "SRR24085752",
                            "single_end": false,
                            "md5_1": "e67296355980997dde170c602d9cb4c1",
                            "md5_2": "86b4ee5d4860aed0a3a2538a7757cbc8"
                        },
                        "SRR24085752.unitigs.fa.gz:md5,134bb7521e3af1d7cf0ceb3c465bdc53"
                    ]
                ],
                "versions": [
                    "versions.yml:md5,63c274313e2b305b08175b44dea333a7"
                ]
            }
        ],
        "meta": {
            "nf-test": "0.9.0",
            "nextflow": "24.04.4"
        },
        "timestamp": "2024-10-10T12:21:53.048141703"
    },
    "stub - filtered + multiplied": {
        "content": [
            {
                "0": [
                    [
                        {
                            "id": "SRR24085752",
                            "single_end": false,
                            "md5_1": "e67296355980997dde170c602d9cb4c1",
                            "md5_2": "86b4ee5d4860aed0a3a2538a7757cbc8"
                        },
                        "SRR24085752.unitigs.fa.gz:md5,68b329da9893e34099c7d8ad5cb9c940"
                    ]
                ],
                "1": [
                    
                ],
                "2": [
                    [
                        {
                            "id": "SRR24085752",
                            "single_end": false,
                            "md5_1": "e67296355980997dde170c602d9cb4c1",
                            "md5_2": "86b4ee5d4860aed0a3a2538a7757cbc8"
                        },
                        "SRR24085752.multiplier.fasta.gz:md5,68b329da9893e34099c7d8ad5cb9c940"
                    ]
                ],
                "3": [
                    [
                        {
                            "id": "SRR24085752",
                            "single_end": false,
                            "md5_1": "e67296355980997dde170c602d9cb4c1",
                            "md5_2": "86b4ee5d4860aed0a3a2538a7757cbc8"
                        },
                        "SRR24085752.filt_mult.fasta.gz:md5,68b329da9893e34099c7d8ad5cb9c940"
                    ]
                ],
                "4": [
                    "versions.yml:md5,63c274313e2b305b08175b44dea333a7"
                ],
                "filt_mult_fasta": [
                    [
                        {
                            "id": "SRR24085752",
                            "single_end": false,
                            "md5_1": "e67296355980997dde170c602d9cb4c1",
                            "md5_2": "86b4ee5d4860aed0a3a2538a7757cbc8"
                        },
                        "SRR24085752.filt_mult.fasta.gz:md5,68b329da9893e34099c7d8ad5cb9c940"
                    ]
                ],
                "filtered_fasta": [
                    
                ],
                "multiplied_fasta": [
                    [
                        {
                            "id": "SRR24085752",
                            "single_end": false,
                            "md5_1": "e67296355980997dde170c602d9cb4c1",
                            "md5_2": "86b4ee5d4860aed0a3a2538a7757cbc8"
                        },
                        "SRR24085752.multiplier.fasta.gz:md5,68b329da9893e34099c7d8ad5cb9c940"
                    ]
                ],
                "raw_fasta": [
                    [
                        {
                            "id": "SRR24085752",
                            "single_end": false,
                            "md5_1": "e67296355980997dde170c602d9cb4c1",
                            "md5_2": "86b4ee5d4860aed0a3a2538a7757cbc8"
                        },
                        "SRR24085752.unitigs.fa.gz:md5,68b329da9893e34099c7d8ad5cb9c940"
                    ]
                ],
                "versions": [
                    "versions.yml:md5,63c274313e2b305b08175b44dea333a7"
                ]
            }
        ],
        "meta": {
            "nf-test": "0.9.0",
            "nextflow": "24.04.4"
        },
        "timestamp": "2024-10-10T12:22:36.272615934"
    },
    "SRR24085752 - filtered": {
        "content": [
            {
                "0": [
                    [
                        {
                            "id": "SRR24085752",
                            "single_end": false,
                            "md5_1": "e67296355980997dde170c602d9cb4c1",
                            "md5_2": "86b4ee5d4860aed0a3a2538a7757cbc8"
                        },
                        "SRR24085752.unitigs.fa.gz:md5,134bb7521e3af1d7cf0ceb3c465bdc53"
                    ]
                ],
                "1": [
                    [
                        {
                            "id": "SRR24085752",
                            "single_end": false,
                            "md5_1": "e67296355980997dde170c602d9cb4c1",
                            "md5_2": "86b4ee5d4860aed0a3a2538a7757cbc8"
                        },
                        "SRR24085752.filter.fasta.gz:md5,5eb2cc6f1a1070e4188c72d53c3c76ae"
                    ]
                ],
                "2": [
                    
                ],
                "3": [
                    
                ],
                "4": [
                    "versions.yml:md5,63c274313e2b305b08175b44dea333a7"
                ],
                "filt_mult_fasta": [
                    
                ],
                "filtered_fasta": [
                    [
                        {
                            "id": "SRR24085752",
                            "single_end": false,
                            "md5_1": "e67296355980997dde170c602d9cb4c1",
                            "md5_2": "86b4ee5d4860aed0a3a2538a7757cbc8"
                        },
                        "SRR24085752.filter.fasta.gz:md5,5eb2cc6f1a1070e4188c72d53c3c76ae"
                    ]
                ],
                "multiplied_fasta": [
                    
                ],
                "raw_fasta": [
                    [
                        {
                            "id": "SRR24085752",
                            "single_end": false,
                            "md5_1": "e67296355980997dde170c602d9cb4c1",
                            "md5_2": "86b4ee5d4860aed0a3a2538a7757cbc8"
                        },
                        "SRR24085752.unitigs.fa.gz:md5,134bb7521e3af1d7cf0ceb3c465bdc53"
                    ]
                ],
                "versions": [
                    "versions.yml:md5,63c274313e2b305b08175b44dea333a7"
                ]
            }
        ],
        "meta": {
            "nf-test": "0.9.0",
            "nextflow": "24.04.4"
        },
        "timestamp": "2024-10-10T12:21:00.454373966"
    }
}